### PR TITLE
Merge ASE feature branch feature/ase_plugin3 into master 

### PR DESCRIPTION
**Pull request by lzhan55**
…… (#1263)

* cherry-pick changes from feature/ase_plugin to feature/ase_plugin2 (#1252)

* ASE: adding ASE as a plugin.
* Lzhan55/ase plugin (#1236)
* Lzhan55/ase plugin2 (#1247)
* ASE: Added a libopae-c-ase.so to make ASE application backward compat… (#1250)

* Added back the missing ASE Makefile

### CODE REVIEW

File: libopae/plugins/ase/CMakeLists.txt [hunk 1]
```diff
@@ -75,12 +75,17 @@ add_subdirectory(rtl)
...
-  ase_regress.sh)
+  ase_regress.sh
+  opae_ase.cfg)
 
 install(FILES ${EXTRA_ASE_FILES}
   DESTINATION ${ASE_INST_SHARE_DIR}
   COMPONENT aseextra)
 
+install(FILES opae_ase.cfg
+  DESTINATION $ENV{HOME}/.local/opae
```
Comment (nakulkorde) [inline libopae/plugins/ase/CMakeLists.txt]: can this file not be installed in the home directory? is there a specific need for this file to be in the home directory? it creates problems when specifying file paths in the rpm specfile. @tswhison what do you think?

Comment (nakulkorde) [reply libopae/plugins/ase/CMakeLists.txt]: I have updated the rpm specfile to incorporate the new changes but I wasn't able to get the home directory to be identified by the specfile.

Comment (tswhison) [reply libopae/plugins/ase/CMakeLists.txt]: Ooh, I can see how the home directory would be problematic for the rpm spec file.  It's probably better that it go into /etc/opae for that reason alone.

Comment (michael-adler) [reply libopae/plugins/ase/CMakeLists.txt]: It's installed in <install root>/share/opae/ase/ too. Is that sufficient and you can skip both .local and /etc?

**Comment by michael-adler** [comment]
Sigh. It appears that gcc 4.5 and later don't work with our new libopae-c-ase strategy. They changed the default in the linker to -no-copy-dt-needed-entries, which causes links against just libopae-c-ase to fail because the functions now in libopae-c are not found. You can try this yourself using a simple example, like samples/tutorial/01_hello_world/sw in intel-fpga-bbb. With gcc 4.5 and later, the cci_hello_ase link fails. If you change the link step, instead, to:

`gcc -o cci_hello_ase obj/cci_hello.o -z noexecstack -z relro -z now -pie -luuid -Wl,-copy-dt-needed-entries -lopae-c-ase`

it works fine. Unfortunately, the whole point of this effort was to be transparent, so forcing a change in linker flags or library lists would defeat the purpose.

The only solution I can think if, so far, is to build libopae-c-ase from the same sources as libopae-c, adding the thunk you have now that puts the run in ASE mode. You would then remove the dependence of libopae-c-ase on libopae-c. It isn't ideal, but it would be easy to do and it would work.

Any better ideas?

File: libopae/plugins/ase/CMakeLists.txt (hunk 1 above)
Comment (lzhan55) [reply libopae/plugins/ase/CMakeLists.txt]: Sure, I will remove it from $HOME.

**Comment by lzhan55** [comment]
@michael-adler Agree with you.

**Comment by michael-adler** [comment]
@deepakcu84 - Have you tried this branch on DCP regressions?

**Comment by deepakcu84** [comment]
well, sorry no. I'll try it.

**Comment by michael-adler** [comment]
You should wait until tracy fixes the problem I noted this morning, described above.

**Comment by deepakcu84** [comment]
Got it. Sorry, I wasn't paying attention.

**Comment by tswhison** [comment]
The coverage number decreased a bit.  Could you add some unit tests for find_ase_cfg( ) and the new init code?

**Comment by lzhan55** [comment]
@tswhison Will add unit tests for that.

**Comment by michael-adler** [comment]
@deepakcu84 - I think it's a good time to test now.

**Comment by michael-adler** [comment]
What’s the plan for this merge? It is starting to get in the way of other changes.

**Comment by lzhan55** [comment]
@michael-adler Amir will take over the ASE work. It could not be merged into master because DCP regression tests have not been run yet. Deepak is working on running the DCP regression tests.

**Comment by lzhan55** [comment]
@michael-adler Thanks for running the DCP regression tests! It is great to see that the DCP regression tests passed with ASE plugins changes.

@nakulkorde Was Amir added as ASE code owner? If yes, Amir could approve the pull requests as ASE owner.

**Comment by deepakcu84** [comment]
Thank you Michael.

**Pull request merged**